bika/lims/browser/worksheet/views/analyses.py

```diff
@@ -71,216 +71,372 @@ def __init__(self, context, request):
-            if obj.portal_type == 'DuplicateAnalysis':
-                pos_text += "<img title='%s' src='%s/++resource++bika.lims.images/duplicate.png'/>" % (_("Duplicate").encode('utf-8'), self.context.absolute_url())
-                pos_text += "<br/>"
-            elif obj.portal_type == 'ReferenceAnalysis' and obj.ReferenceType == 'b':
-                pos_text += "<a href='%s'><img title='%s' src='++resource++bika.lims.images/blank.png'></a>" % (parent.absolute_url(), parent.Title())
-                pos_text += "<br/>"
-            elif obj.portal_type == 'ReferenceAnalysis' and obj.ReferenceType == 'c':
-                pos_text += "<a href='%s'><img title='%s' src='++resource++bika.lims.images/control.png'></a>" % (parent.absolute_url(), parent.Title())
-                pos_text += "<br/>"
-            if parent.portal_type == 'AnalysisRequest':
-                sample = parent.getSample()
-                pos_text += "<a href='%s'><img title='%s' src='++resource++bika.lims.images/sample.png'></a>" % (sample.absolute_url(), sample.Title())
-            pos_text += "</td></tr>"
-
-            pos_text += "<tr><td>"
-            if parent.portal_type == 'AnalysisRequest':
-                pos_text += "<a href='%s'>%s</a>" % (parent.absolute_url(), parent.Title())
-            elif parent.portal_type == 'ReferenceSample':
-                pos_text += "<a href='%s'>%s</a>" % (parent.absolute_url(), parent.Title())
-            elif obj.portal_type == 'DuplicateAnalysis':
-                pos_text += "<a style='white-space:nowrap' href='%s'>%s</a>" % (obj.getAnalysis().aq_parent.absolute_url(), obj.getReferenceAnalysesGroupID())
-            elif parent.portal_type == 'Worksheet':
-                parent = obj.getAnalysis().aq_parent
-                pos_text += "<a href='%s'>(%s)</a>" % (parent.absolute_url(), parent.Title())
-            pos_text += "</td></tr>"
-
-            # sampletype
-            pos_text += "<tr><td>"
-            if obj.portal_type == 'Analysis':
-                pos_text += obj.aq_parent.getSample().getSampleType().Title()
-            elif obj.portal_type == 'ReferenceAnalysis' or \
+            # This is the first analysis found for the given position, add the
+            # slot info in there and apply a rowspan accordingly.
+            rowspan = self.items_rowspans.get(item_position, 1)
+            prev_position = item_position
+            item['rowspan'] = {'Pos': rowspan}
+            item['replace']['Pos'] = self.get_slot_header(item)
+
+    def get_slot_header(self, item):
+        """
+        Generates a slot header (the first cell of the row) for the item
+        :param item: the item for which the slot header is requested
+        :return: the html contents to be displayed in the first cell of a slot
+        """
+        obj = item['obj']
+        obj = api.get_object(obj)
+
+        # TODO All contents below have to be refactored/cleaned-up!
```

Comment:
Indeed. HTML generation in code always feels awkward to me. Maybe we can factor that out in the next iterations